text/0000-let-else.md

```diff
@@ -0,0 +1,473 @@
+
+One possible benefit of adding a keyword is that it could make a possible future extension for similarity to the (yet unimplemented) [if-let-chains][] feature more straightforward.
+However, as mentioned in the [future-possibilities][] section, this is likely not necessary.
+
+This syntax has prior art in the Swift programming language, which includes a [guard-let-else][swift] statement
+which is roughly equivalent to this proposal except for the choice of keywords.
+
+### `let PATTERN = EXPR else return EXPR;`
+
+A potential alternative to requiring parentheses in `let PATTERN = (if { a } else { b }) else { c };` is to change the syntax of the `else` to no longer be a block
+but instead an expression which starts with a diverging keyword, such as `return` or `break`.
+
+Example:
+```
+let Some(foo) = some_option else return None;
+```
+
+This RFC avoids this because it is overall less consistent with `else` from if-else, which require blocks.
+
+This was originally suggested in the old RFC, comment at https://github.com/rust-lang/rfcs/pull/1303#issuecomment-188526691
+
+### `else`-block fall-back assignment
+
+A fall-back assignment alternate to the diverging block has been proposed multiple times in relation to this feature in the [original rfc][] and also in out-of-RFC discussions.
+
+This RFC avoids this proposal, because there is no clear syntax to use for it which would be consistent with other existing features.
+Also use-cases for having a single fall-back are much more rare and unusual, where as use cases for the diverging block are very common.
+This RFC proposes that most fallback cases are sufficiently or better covered by using `match`.
+
+An example, using a proposal to have the binding be visible and assignable from the `else`-block.
+Note that this is incompatible with this RFC and could probably not be added as an extension from this RFC.
+
+```rust
+enum AnEnum {
+    Variant1(u32),
+    Variant2(String),
+}
+
+let AnEnum::Variant1(a) = x else {
+    a = 42;
+};
+```
+
+Another potential alternative for fall-back which could be added with an additional keyword as a future extension:
+
+```rust
+enum AnEnum {
+    Variant1(u32),
+    Variant2(String),
+}
+
+let AnEnum::Variant1(a) = x else assign a {
+    a = 42;
+};
+```
+
+### `if !let PAT = EXPR { BODY }`
+
+The [old RFC][old-rfc] originally proposed this general feature via some kind of pattern negation as `if !let PAT = EXPR { BODY }`.
+
+This RFC avoids adding any kind of new or special pattern matching rules. The pattern matching works as it does for if-let.
+The general consensus in the old RFC was also that the negation syntax is much less clear than `if PATTERN = EXPR_WITHOUT_BLOCK else { /* diverge */ };`,
+and partway through that RFC's lifecycle it was updated to be similar to this RFC's proposed let-else syntax.
+
+### Complete Alternative
```

Comment:
```suggestion
### Null Alternative
```

Or some other heading. "Complete Alternative" doesn't seem like it conveys the idea of "we could also do nothing".